kclvm/sema/src/advanced_resolver/mod.rs

```diff
@@ -0,0 +1,1038 @@
+  │        │ ast::Expression │                                                   │
+  │        └─────────────────┘                                                   │
+  │                 │                                                            │
+  │                 │ resolve_local_value                                        │
+  │                 ▼                                                            │
+  │        ┌─────────────────┐                                                   │
+  │        │ ast::Expression │                                                   │
+  │        └─────────────────┘                                                   │
+  │                 │                                                            │
+  │                 │ resolve_symbol_ref (map Expression to DefinitionSymbols)   |
+  │                 ▼                                                            │
+  │        ┌─────────────────┐                                                   │
+  │        │ ast::Expression │                                                   │
+  │        └─────────────────┘                                                   │
+  └──────────────────────────────────────────────────────────────────────────────┘
+                                 │
+                                 ▼   build_sema_db (collect symbol locs and analyse scope)
+                        ┌─────────────────────┐
+                        │ core::GlobalState   │
+                        └─────────────────────┘
+  AdvancedResolver mainly does two tasks:
+      1: Traverse AST to parse LocalSymbol and store it in GlobalState, while storing the parsed type in Symbol
+      2: Establish a mapping between expressions and SymbolRef, specifically injecting symbol information into AST
+
+  After the work of the advanced resolver is completed, the GlobalState will build the whole semantic database,
+  so that toolchain can query semantic information about the AST
+
+*/
+
+use std::rc::Rc;
+
+use indexmap::IndexMap;
+use kclvm_error::Position;
+
+use crate::{
+    core::{
+        global_state::GlobalState,
+        package::ModuleInfo,
+        scope::{LocalSymbolScope, RootSymbolScope, ScopeKind, ScopeRef},
+        symbol::SymbolRef,
+    },
+    ty::Type,
+};
+use kclvm_ast::ast;
+use kclvm_ast::ast::AstIndex;
+use kclvm_ast::ast::Program;
+use kclvm_ast::walker::MutSelfTypedResultWalker;
+mod node;
+/// Resolver is responsible for program semantic checking, mainly
+/// including type checking and contract model checking.
+pub struct AdvancedResolver<'ctx> {
+    pub(crate) ctx: Context<'ctx>,
+    pub(crate) gs: GlobalState,
+}
+
+pub struct Context<'ctx> {
+    pub program: &'ctx Program,
+    node_ty: IndexMap<AstIndex, Rc<Type>>,
+    scopes: Vec<ScopeRef>,
+    current_pkgpath: Option<String>,
+    current_filename: Option<String>,
+    current_schema_symbol: Option<SymbolRef>,
+    start_pos: Position,
+    end_pos: Position,
+    maybe_def: bool,
```

Comment:
What does the attribute `maybe_def` means (the left and right values)? Would U pls add more documents?